### ISSUE TITLE
Fourier Features Let Networks Learn High Frequency Functions in Low Dimensional Domains

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#71]
*NIPS 2020 Spotlight*

## Motivation

- MLPs fail to learn high frequency functions in low dimensional domain (ex: coordinate -> RGB)
- Adding sinusoidal positional encoding greatly boosts high frequency region fitting. (Empirical in NeRF)

## Key Idea

- Viewing MLPs in NTK regime, adding Fourier feature mapping not only transforms the kernel into a stationary one, but also provides the capacity to *control* kernel (band) width.
- Kernel width holds the key to convergence rate (whether MLP can learn high frequency region) and degree of overfitting.

## Previous Approach

- **SIREN**: replaces ReLU with sinusoidal activation to express high frequency details.
- **NeRF**: adds Fourier feature mapping to (x,y,z) coordinate queries and significantly improve visual synthesis quality.

### Kernel Regression expresses prediction as weighted sum of training labels

![image](https://user-images.githubusercontent.com/46308407/125880281-3c1b481c-5941-4709-aaae-830d9fb2de4a.png)

### NTK explains training dynamics of neural networks through the lens of kernel regression

![image](https://user-images.githubusercontent.com/46308407/125880347-bb6d0767-8885-4831-a2a4-d1aae199ff61.png)

![image](https://user-images.githubusercontent.com/46308407/125880375-58a8ab5b-99ba-42b7-aa31-8bfa1a3742cd.png)

### Spectral bias of MLPs are caused by rapidly decaying convergence rate

![image](https://user-images.githubusercontent.com/46308407/125880605-973d2560-8adb-47fa-acfc-8d31ffa608c9.png)

- Components of target function corresponding to larger eigenvalues are learned faster, but in conventional MLPs eigenvalues of NTK decay rapidly (resulting in failed convergence especially in higher frequency region)

### NTKs composed from Fourier features

![image](https://user-images.githubusercontent.com/46308407/125880870-a536a4f5-623a-4301-aa4b-78dfca93584e.png)
![image](https://user-images.githubusercontent.com/46308407/125880890-178afc08-5206-4be4-ac81-aad69c656e56.png)
![image](https://user-images.githubusercontent.com/46308407/125880925-1997e716-2d64-4d31-a0fb-899cb8fc8214.png)

- Now it's stationary, and a_j, b_j can control the bandwidth.

![image](https://user-images.githubusercontent.com/46308407/125881045-7253d530-d201-4cc1-b7bf-f795b17b5cfd.png)

- Lower **p** corresponds to narrower kernel bandwidth (more suitable for learning higher frequency but prone to overfitting at the same time)

### Lower p (narrower bandwidth) corresponds to overfitting

![image](https://user-images.githubusercontent.com/46308407/125881188-db2c38cc-e1e9-4599-a92e-f26bbf02666a.png)

### In practice, just sample Fourier coefficients from any distribution with carefully chosen variance.
![image](https://user-images.githubusercontent.com/46308407/125881241-530bc7c7-fb66-4b4e-a0d7-32547f5f07c4.png)

- It's variance that matters, not the distribution (shape) itself.

### Quantitative Evaluations

![image](https://user-images.githubusercontent.com/46308407/125881334-eddc6e03-dd66-45e8-ac6c-81d499c526a5.png)

### Qualitative Evaluations

![image](https://user-images.githubusercontent.com/46308407/125881366-d62f914e-d0d2-49f2-99d1-03ea36b8b499.png)

## Comments

- Fourier transformations are widely explored these days (FNet, SIREN, Transformer position embedding, etc)
- I've always had doubts about the usefulness of NTKs, but in terms of interpreting deep learning dynamics it can be extremely powerful.
- I've never thought of low dim dense domain, nor realize that usual image data we are dealing with belong to high dim sparse domain. It was a revelation indeed.
